Restrict the use of `any` type

chidexebere opened the pull request:
## Changes

- This adds the`"@typescript-eslint/no-explicit-any": "error"` type rule to eslint (`eslintrc.json`) to prevent the use of the `any` type.
- adds `"lint": "eslint . --ext .js,.ts,.tsx"` script to `package.json` to run lint
- adds `public` to `eslintignore`
- Fixes existing `any` type errors and the other type errors

## Checklist

- [x] I have checked out our [styleguide](https://github.com/PostHog/posthog.com/blob/master/STYLEGUIDE.md)

mariusandra reviewed (changes requested):
I appreciate the effort, and while we have a bunch of lazy `any`s that just need to be replaced with real types, I'm not sure replacing the other `any`s with `unknown`s and `@eslint` exceptions is a great improvement in code quality.

Thus I'm in favour of not being that strict.

chidexebere commented:
@mariusandra  If you look through my changes, I did not just change `any` type to `unknown` type; I did replace with the actual types. I only used `unknown` type in few places where the actual type was not known. Also I only used eslint exceptions as a last resort in also few places after testing with the `unknown` type and doing a type check.
 
 If you look through your codebase, you will notice that you just dont have a bunch of `any` types but a lot of them. You can check how many `children: any` I changed to `children: React.ReactNode`. I understand the need to build things faster but this is building up and already causing issues as mentioned by @yakkomajuri  in #1333.
 
 The essence of the `any` type is for easy migration from JS to TS and not for use any time a developer does not feel like checking for the right type.
 
 Although this is my suggestion on code improvement, it is still left for the team to decide.

mariusandra reviewed (changes requested):
@chidexebere I did look through your changes before making my comment. 

With `unknown`, I was referring to this case: `(posts: Array<unknown>)`. Posts aren't `unknown` here. They're just not typed. So `any` is a more accurate type.

Also, again I don't see a difference between allowing `any`s and disallowing `any`s, yet still using them and adding `eslint-disable-next-line` exceptions around them. I think adding this strict eslint rule actually hurts us more.

I get the issue you're trying to solve. I'd love a beautiful, fully typed, and strict codebase, but considering:

1) we'd still be using `any`s, just differently
2) this will inevitably hurt new contributions to the codebase
3) at the end of the day this is a static website, not a complicated SaaS app, so we don't need to be that strict

I'd ask you to make some changes. Namely, can we just replace "error" with "warn" in eslintrc?

mariusandra commented:
Also, @yakkomajuri, feel free to share your thoughts and or merge this as is. It's fine at the end of the day. I'm also not going to be the one working inside this codebase, so... :).

yakkomajuri commented:
Hey @chidexebere! Thanks for this and sorry for the delay.

I'm with @mariusandra here. If we could change from an error to a warning and type some of the stuff the we actually know the types for (instead of using `unknown`) that'd be great.

chidexebere commented:
> Hey @chidexebere! Thanks for this and sorry for the delay.
> 
> I'm with @mariusandra here. If we could change from an error to a warning and type some of the stuff the we actually know the types for (instead of using `unknown`) that'd be great.

noted, I will make the changes.

leggetter commented:
@chidexebere - this has been open for a while and I'm looking to do some tidying up. Do you feel you've addressed the points raised by @mariusandra and @yakkomajuri ?

chidexebere commented:
> @chidexebere - this has been open for a while and I'm looking to do some tidying up. Do you feel you've addressed the points raised by @mariusandra and @yakkomajuri ?

@leggetter , Yes I have. This can be seen from my last commits

Twixes reviewed (approved):
This looks really nice now, thank you @chidexebere
